### PR TITLE
feat(split): child replica apply private logs, in-memory mutations and catch up parent

### PR DESCRIPTION
**Pull request by hycdong**
## Simple partition split process
1. meta receives client partition split request, and change partition count #286
2. replica notices partition count changed during on_config_sync
3. parent partition create child partition #291
4. parent prepare states for child to learn #299 
5. child partition async learn states from parent
    - copy prepare list #309 
    - apply checkpoint #309 
    - **replay private logs**
    - **learn in-memory mutations**
    - **catch up parent states**
6. all child partitions in group finish learn process, parent will stop read and write
7. meta server register child partitions
8. child partition active, and parent recover read and write

More partition split discussion in issue #69 and [partition split design doc](https://pegasus-kv.github.io/2020/02/06/partition-split-design.html)
This pr solves the part of fifth step of partition split, which is bold in process description.

## What this pr solved
- `child_apply_private_logs`
    - child apply private logs and in-memory mutations
    - similar to function `apply_learned_state_from_private_log` during learn
- `child_catch_up_states`
    - similar to function `init_learn` with learner_status `LearningWithPrepare`
    - child will also prepare mutations from parent (which will show in further pr) during async-learn, as a result, child prepare_list last_committed_decree should be equal to parent prepare list last_committed_decree, which is the `goal_decree`, and `local_decree` is child local last_committed_decree which is the last decree in async-learn.
    - if `local_decree < goal_decree`, it means there are mutations written to parent during async-learn, child does not catch up parent, there are still some mutations child not learn.
    - if `local_decree >= _prepare_list->min_decree()`, it means all missing mutations are all in prepare list, otherwise, some missing mutations have already in private log, should call `catch_up_with_private_logs` to catch up all missing mutations
- add `is_caught_up` flag for partition_split context to identify if child catch up parent

## Tests
- add replay log failed case for `child_learn_states`
- add unit tests for `child_apply_private_logs` and `child_catch_up`

### CODE REVIEW

File: src/dist/replication/lib/replica_split.cpp [hunk 1]
```diff
@@ -330,30 +325,147 @@ void replica::child_learn_states(learn_state lstate,
...
+            for (decree d = local_decree + 1; d <= goal_decree; ++d) {
+                auto mu = _prepare_list->get_mutation_by_decree(d);
+                dassert(mu != nullptr, "");
+                error_code ec = _app->apply_mutation(mu);
+                if (ec != ERR_OK) {
+                    child_handle_split_error("child_catchup failed because apply mutation failed");
+                    return;
+                }
+            }
+        } else { // some missing mutations are not in memory
```
Comment (levy5307) [inline src/dist/replication/lib/replica_split.cpp]: 这个整个流程就是，先apply private log, 然后apply in-memory mutation，最后如果apply本地的prepare-list，我理解的没错吧？这里的else的情况就是还没有apply到本地prepare-list，然后再继续等待，直到开始apply本地prepare-list，是吧？

File: src/dist/replication/lib/replica_context.h [hunk 2]
```diff
@@ -522,13 +522,14 @@ typedef dsn::ref_ptr<cold_backup_context> cold_backup_context_ptr;
...
 {
 public:
-    partition_split_context() : is_prepare_list_copied(false) {}
+    partition_split_context() : is_prepare_list_copied(false), is_caught_up(false) {}
     bool cleanup(bool force);
     bool is_cleaned() const;
 
 public:
     gpid parent_gpid;
     bool is_prepare_list_copied;
```
[the remark this answers is not in the document]
Comment (hycdong) [reply src/dist/replication/lib/replica_context.h]: 这个变量后面的pr会get

File: src/dist/replication/lib/replica_split.cpp [hunk 3]
```diff
@@ -95,6 +95,7 @@ void replica::child_init_replica(gpid parent_gpid,
     // init split states
     _split_states.parent_gpid = parent_gpid;
```
[the remark this answers is not in the document]
Comment (hycdong) [reply src/dist/replication/lib/replica_split.cpp]: _split_states还有其他变量

File: src/dist/replication/lib/replica_split.cpp [hunk 4]
```diff
@@ -330,30 +325,147 @@ void replica::child_learn_states(learn_state lstate,
...
+
+    if (status() != partition_status::PS_PARTITION_SPLIT) {
+        derror_replica("wrong status = {}", status());
+        return;
+    }
+
+    decree goal_decree = _prepare_list->last_committed_decree();
+    decree local_decree = _app->last_committed_decree();
+
+    // there are still some mutations child not learn
```
[the remark this answers is not in the document]
Comment (hycdong) [reply src/dist/replication/lib/replica_split.cpp]: 添加了对`child_catch_up_states`的说明

File: src/dist/replication/test/replica_test/unit_test/replica_split_test.cpp [hunk 5]
```diff
@@ -151,12 +167,34 @@ class replica_split_test : public testing::Test
...
     void test_child_learn_states()
     {
-        mock_child_async_learn_states(_child, true);
+        mock_child_async_learn_states(_child, true, _decree);
         _child->child_learn_states(
             _mock_learn_state, _mutation_list, _private_log_files, _total_file_size, _decree);
         _child->tracker()->wait_outstanding_tasks();
     }
 
+    void test_child_apply_private_logs()
```
[the remark this answers is not in the document]
Comment (hycdong) [reply src/dist/replication/test/replica_test/unit_test/replica_split_test.cpp]: 没其他地方用了

File: src/dist/replication/lib/replica_split.cpp [hunk 6]
```diff
@@ -330,30 +325,147 @@ void replica::child_learn_states(learn_state lstate,
...
 {
     FAIL_POINT_INJECT_F("replica_child_catch_up_states", [](dsn::string_view) {});
-    // TODO(heyuchen): TBD
+
+    if (status() != partition_status::PS_PARTITION_SPLIT) {
+        derror_replica("wrong status = {}", status());
+        return;
+    }
+
+    decree goal_decree = _prepare_list->last_committed_decree();
```
[the remark this answers is not in the document]
Comment (hycdong) [reply src/dist/replication/lib/replica_split.cpp]: 添加了对`child_catch_up_states`的说明

File: src/dist/replication/lib/replica_split.cpp [hunk 7]
```diff
@@ -330,29 +325,155 @@ void replica::child_learn_states(learn_state lstate,
...
+            dwarn_replica(
+                "there are some private logs should be learned, app last_committed_decree="
+                "{}, prepare_list min_decree={}, please wait",
+                local_decree,
+                _prepare_list->min_decree());
+            _split_states.async_learn_task = tasking::enqueue(
+                LPC_CATCHUP_WITH_PRIVATE_LOGS,
+                tracker(),
+                [this]() {
+                    catch_up_with_private_logs(partition_status::PS_PARTITION_SPLIT);
```
Comment (neverchanje) [inline src/dist/replication/lib/replica_split.cpp]: catch_up_with_private_logs 后续 PR 建议改个名字，这个函数和 child_apply_private_logs 对代码不熟悉的很容易看混，而且不明其意

File: src/dist/replication/lib/replica_context.h [hunk 8]
```diff
@@ -522,13 +522,16 @@ typedef dsn::ref_ptr<cold_backup_context> cold_backup_context_ptr;
...
-    partition_split_context() : is_prepare_list_copied(false) {}
+    partition_split_context() : is_prepare_list_copied(false), is_caught_up(false) {}
     bool cleanup(bool force);
     bool is_cleaned() const;
 
 public:
     gpid parent_gpid;
-    bool is_prepare_list_copied;
+    // whether child has copied parent prepare list
+    bool is_prepare_list_copied{false};
```
Comment (acelyc111) [inline src/dist/replication/lib/replica_context.h]: 使用了这种形式初始化，构造函数就不用再初始化了

**Pull request merged**